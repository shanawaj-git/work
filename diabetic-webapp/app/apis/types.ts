export type OTPGenerationResponse = {
  success: boolean;
};
export type PrescriptionUploadResponse = {
  id: number;
};

export type AuthUser = {
  id: number;
  username: string;
  email: string;
  blocked: false;
  firstName: string;
  middleName: string;
  lastName: string;
  mobileNumber: string;
  emiratesId: string;
  reminderFrequency: number;
  createdAt: string;
  updatedAt: string;
};

export type OTPValidationResponse = {
  jwt: string;
  user: AuthUser;
};

export type APIClientOptions = {
  onSuccessfulAuthentication: (response: OTPValidationResponse) => void;
  onUnauthorizedAccess: () => void;
  loginUrl: string;
};

export enum AuthDataKeys {
  Jwt = 'auth_jwt',
  User = 'auth_user',
}

export type OrdersInitializeResponse = {
  id: number;
  schedule: string;
  amount: number;
  currency: string;
  status: string;
  address: Address;
  prescription: Prescription;
};

export type Address = {
  id?: number;
  flatOrVillaNumber: number;
  buildingOrVilla: string;
  street: string;
  area?: string;
  directions?: string;
  latitude: number;
  longitude: number;
};

export type Prescription = {
  id: number;
  firstName: string;
  lastName: string;
  status: string;
};

export type OrdersInitializeRequest = {
  prescriptionNumber: string;
  address: Address;
  schedule: string;
};

export enum OrderStatus {
  Pending = 'Pending',
  InsuranceApproved = 'Insurance Approved',
  RequirePayment = 'Require Payment',
  PaymentReceived = 'Payment Received',
  DeliveryScheduled = 'Delivery Scheduled',
  Fulfilled = 'Fulfilled',
  Failed = 'Failed',
  Canceled = 'Canceled',
}

export enum PaymentSubType {
  Mastercard = 'Mastercard',
  Visa = 'Visa',
  ApplePay = 'Apple Pay',
  GooglePay = 'Google Pay',
  Cash = 'Cash',
}
export enum PaymentType {
  Card = 'Card',
  Wallet = 'Wallet',
  Cash = 'Cash',
  CashOnDelivery = 'Cash On Delivery',
  CardOnDelivery = 'Card On Delivery',
}

export enum PaymentStatus {
  Created = 'Created',
  Processing = 'Processing',
  PaymentMethodRequired = 'Payment Method Required',
  ActionRequired = 'ActionRequired',
  Canceled = 'Canceled',
  CaptureRequired = 'CaptureRequired',
  Captured = 'Captured',
  PartiallyCaptured = 'PartiallyCaptured',
  Succeeded = 'Succeeded',
  Failed = 'Failed',
  Unknown = 'Unknown',
}

export enum Currency {
  AED = 'AED',
}

export type Payment = {
  id: number;
  paymentType: PaymentType;
  paymentSubType: PaymentSubType;
  status: string;
  maskedCardNumber: string;
  providerMetadata: object;
  createdAt: string;
};

export type Order = {
  id: number;
  amount: number;
  currency: Currency;
  status: OrderStatus;
  payment: Payment;
};

export type Pharmacy = {
  pharmacyName: string;
  supportContactNumber: string;
  scheduleWindowDays: number;
  deliveryHours: object;
};
