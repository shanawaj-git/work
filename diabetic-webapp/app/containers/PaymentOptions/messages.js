import { defineMessages } from 'react-intl';

const scope = 'orders.payment.input';

export default defineMessages({
  headerPrefix: {
    id: `${scope}.header.prefix`,
    defaultMessage: 'Select your',
  },

  headerSuffix: {
    id: `${scope}.header.suffix`,
    defaultMessage: 'payment method',
  },

  amountLabel: {
    id: `${scope}.amount.label`,
    defaultMessage: 'Total to pay',
  },

  onlinePayment: {
    id: `${scope}.option.online`,
    defaultMessage: 'Pay online',
  },

  cashOnDelivery: {
    id: `${scope}.option.cash.delivery`,
    defaultMessage: 'Cash on delivery',
  },

  cardOnDelivery: {
    id: `${scope}.option.card.delivery`,
    defaultMessage: 'Card on delivery',
  },
});
