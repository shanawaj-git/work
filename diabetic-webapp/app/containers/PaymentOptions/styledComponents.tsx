import styled from 'styled-components';
import { colorSelector } from 'themes';
import { Typography } from '@albathanext/design-system';

export const MainContainer = styled.div`
  display: flex;
  background-color: ${colorSelector('p400')};
  flex: 1 1 auto;
  flex-direction: column;
  overflow-y: auto;
  padding: 24px;
  max-width: 600px;
  height: 100%;
  padding-top: 22px;
  margin: auto;
  gap: 8px;
`;
export const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
`;

export const MpcLogo = styled.img`
  display: flex;
  width: 54px;
  height: 45px;
  align-self: flex-end;
`;

export const TitleContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding-bottom: 16px;
  align-items: baseline;
`;

export const TitlePrefix = styled(Typography)`
  &&.MuiTypography-root {
    font-weight: 400;
    font-size: 32px;
    line-height: 36px;
    font-style: normal;
  }
`;

export const TitleSuffix = styled(Typography)`
  &&.MuiTypography-root {
    font-weight: 600;
    font-size: 32px;
    line-height: 36px;
    font-style: normal;
  }
`;

export const AmountContainer = styled.div`
  display: flex;
  padding-bottom: 16px;
  align-items: baseline;
  justify-content: space-between;
  margin-top: 24px;
`;

export const AmountLabel = styled(Typography)`
  &&.MuiTypography-root {
    font-style: normal;
    font-weight: 600;
    font-size: 15px;
    line-height: 22px;
  }
`;

export const AmountValue = styled(Typography)`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 4px 12px;
  background: #020a0a;
  border-radius: 8px;
  &&.MuiTypography-root {
    color: #ffffff;
    text-transform: uppercase;
    font-style: normal;
    font-weight: 600;
    font-size: 15px;
    line-height: 22px;
    text-align: center;
    letter-spacing: 0.3px;
  }
`;

export const OptionsContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  gap: 12px;
  padding-bottom: 80px;
  margin-top: auto;
  .MuiButton-root {
    background: #ffffff;
    border-radius: 8px;
  }
  > button {
    width: 100%;
    height: 60px;
    > div {
      width: 100%;
      height: 100%;
    }
  }
  .MuiButton-root:hover {
    .MuiTypography-root {
      color: #ffffff;
    }
  }
`;

export const OptionContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin: auto;
  height: 100%;
  padding: 16px;

  .MuiTypography-root {
    font-family: 'Neue Haas Grotesk Display Pro';
    font-style: normal;
    font-weight: 400;
    font-size: 15px;
    line-height: 22px;
    letter-spacing: 0.3px;
    color: ${colorSelector('p500')};
  }

  .option-icon {
    height: 24px;
    width: 26px;
  }
`;

export const ButtonLabel = styled(Typography)``;
