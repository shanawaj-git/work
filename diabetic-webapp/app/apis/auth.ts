import { getAPIClientInstance } from './client';
import { OTPGenerationResponse, OTPValidationResponse } from './types';

const Urls = {
  GENERATE_OTP: '/api/users-permissions/otp',
  VALIDATE_OTP: '/api/auth/local',
};

enum Errors {
  UnknownUser='auth.user.unknown',
  InvalidOTP='auth.otp.invalid',
  ExpiredOTP='auth.otp.expired',
  WaitToRegenerateOTP='auth.error.otp.wait',
  MaxFailedAttempt='auth.error.otp.limit'
}

export const publicEndpointConfig = {
  isPublicUrl: true,
}

const generateOTP = async (mobileNumber: string): Promise<OTPGenerationResponse> => {
  try {
    const response = await getAPIClientInstance().post(Urls.GENERATE_OTP, {
      mobileNumber,
    }, publicEndpointConfig);

    return { success: response.data.success };
  } catch (err) {
    if (err.response?.status === 404) {
      throw new Error(Errors.UnknownUser);
    } else if (err.response?.status === 403) {
      if (err.response.data.error.details.errorKey === Errors.WaitToRegenerateOTP)
        throw new Error(Errors.WaitToRegenerateOTP);
    } else {
      throw err;
    }
  }
};

const validateOTP = async (mobileNumber: string, otp: string): Promise<OTPValidationResponse> => {
  try {
    const { data: { jwt, user } } = await getAPIClientInstance().post(Urls.VALIDATE_OTP, {
      identifier: mobileNumber,
      password: otp,
    }, publicEndpointConfig);

    return { jwt, user: {...user} };
  } catch (err) {
    const { response: { status = undefined } = {}} =  err;
    if (status === 400) {
      throw new Error(Errors.InvalidOTP);
    } else if (status === 403) {
      if (err.response.data.error.details.errorKey === Errors.MaxFailedAttempt)
        throw new Error(Errors.MaxFailedAttempt);
      if (err.response.data.error.details.errorKey === Errors.ExpiredOTP)
        throw new Error(Errors.ExpiredOTP);
    } else {
      throw err;
    }
  }
};

export {
  generateOTP,
  validateOTP,
  Urls as AuthApiUrls,
  Errors as AuthApiErrors,
};
