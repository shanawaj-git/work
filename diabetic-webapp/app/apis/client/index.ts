import { APIClientOptions } from 'apis/types';
import axios from 'axios';
import { addInterceptors } from './addInterceptors';
import * as environmentConfig from '../../configs/environmentConfig';

let httpClient;
// eslint-disable-next-line no-unused-vars
const configureAPIClient = (options: APIClientOptions) => {
  // TODO: intercept response and add callback support for login, auth failed
  httpClient = axios.create({
    baseURL:
      environmentConfig.getEnvironmentConfig().API_BASE_URL ||
      'http://localhost:1337',
    timeout: 30000, // 30 seconds
    // TODO: pass Authorization once login is implemented
  });
  addInterceptors(httpClient, options);
  return httpClient;
};

const getAPIClientInstance = () => {
  if (!httpClient) {
    throw new Error('API Client is not intialized');
  }
  return httpClient;
};

export { configureAPIClient, getAPIClientInstance, APIClientOptions };
