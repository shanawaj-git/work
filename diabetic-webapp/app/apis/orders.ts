import { getAPIClientInstance } from './client';
import {
  OrdersInitializeResponse,
  OrdersInitializeRequest,
  Order,
  PaymentType,
} from './types';

const Urls = {
  ORDERS_INITIALIZE: '/api/orders/initialize',
  orderDetails: (id: string) => `/api/orders/${id}`,
  confirmPayment: (id: string) => `/api/orders/${id}/payments/confirmation`,
};

enum Errors {
  PrescriptionNotFound = 'order.prescription.not.found',
  OrderAlreadyExists = 'order.already.exits',
}

const initializeOrder = async (
  orderInitializeRequest: OrdersInitializeRequest,
): Promise<OrdersInitializeResponse> => {
  try {
    return (await getAPIClientInstance().post(
      Urls.ORDERS_INITIALIZE,
      orderInitializeRequest,
    )).data;
  } catch (error) {
    const { response: { status = undefined } = {} } = error;
    if (status === 404) {
      throw new Error(Errors.PrescriptionNotFound);
    } else if (status === 409) {
      throw new Error(Errors.OrderAlreadyExists);
    } else {
      throw error;
    }
  }
};

const fetchOrderDetails = async (orderId: string): Promise<Order> => {
  return (await getAPIClientInstance().get(Urls.orderDetails(orderId))).data;
};

const confirmPayment = async (
  orderId: string,
  type?: PaymentType,
): Promise<Order> => {
  const args: [any] = [Urls.confirmPayment(orderId)];
  if(type) {
    args.push({
      paymentMethod: type,
    });
  }
  return (await getAPIClientInstance().put(...args)).data;
};

export {
  initializeOrder,
  fetchOrderDetails,
  confirmPayment,
  Urls as OrdersApiUrls,
  Errors as OrderApiErrors,
};
