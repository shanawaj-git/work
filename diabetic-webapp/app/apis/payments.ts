import { getAPIClientInstance } from './client';
import { Payment } from './types';

export const Urls = {
  PAYMENTS_INITIALIZE: '/api/payments/initialize',
};

export const initializePayment = async (orderId: string): Promise<Payment> =>
  (await getAPIClientInstance().post(Urls.PAYMENTS_INITIALIZE, {
    orderId,
  })).data;
