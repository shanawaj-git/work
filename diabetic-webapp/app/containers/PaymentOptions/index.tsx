import React, { useEffect, useState } from 'react';
import { injectIntl, intlShape } from 'react-intl';
import PropTypes, { InferProps } from 'prop-types';
import { useParams } from 'react-router-dom';
import { Button } from '@albathanext/design-system';

import Loader from 'components/moclecules/Loader';
import { Icon } from '@mui/material';
import * as ordersApi from '../../apis/orders';

import { Order, OrderStatus, PaymentType } from '../../apis/types';
import {
  TitleContainer,
  LoaderContainer,
  MainContainer,
  MpcLogo,
  TitlePrefix,
  TitleSuffix,
  AmountContainer,
  AmountLabel,
  AmountValue,
  OptionsContainer,
  ButtonLabel,
  OptionContainer,
} from './styledComponents';
import messages from './messages';
const mpcGrayLogo = require('images/mpc_gray.svg');
const logoOnlinePayment = require('images/online-payment.svg');
const logoCashOnDelivery = require('images/cash-on-delivery.svg');
const logoCardOnDelivery = require('images/card-on-delivery.svg');

const PAYMENT_STATUS_PAGE = 'payments/status';
const PAYMENT_INPUT_PAGE = 'payments/input';

const validOrderStatuses = [
  OrderStatus.InsuranceApproved,
  OrderStatus.RequirePayment,
];

export const SEARCH_PARAM_CONFIRMATION = 'confirmation';

const ProgressBar = () => (
  <LoaderContainer>
    <Loader size={90} />
  </LoaderContainer>
);

const Option = ({ intl, text, testid, onClick, icon }: any) => (
  <Button
    data-testid={testid}
    size="large"
    variant="contained"
    onClick={onClick}
  >
    <OptionContainer>
      <ButtonLabel>{intl.formatMessage(text)}</ButtonLabel>
      <Icon>
        <img className="option-icon" alt={text} src={icon} />
      </Icon>
    </OptionContainer>
  </Button>
);

const Options = ({
  order,
  intl,
  history,
}: {
  order: Order;
  intl: any;
  history: any;
}) => {
  const [processing, setProcessing] = useState(false);

  const handleOnlinePayment = () => history.push(PAYMENT_INPUT_PAGE);
  const redirectToOrderStatus = () => history.push(PAYMENT_STATUS_PAGE);

  const confirmPayment = async (type: PaymentType) => {
    try {
      setProcessing(true);
      await ordersApi.confirmPayment(`${order.id}`, type);
    } catch (err) {
      // In case of error too it should redirect to the status page.
    } finally {
      setProcessing(false);
      redirectToOrderStatus();
    }
  };

  const handleCashOnDelivery = () => confirmPayment(PaymentType.CashOnDelivery);
  const handleCardOnDelivery = () => confirmPayment(PaymentType.CardOnDelivery);


  return (
    <>
      {processing && <ProgressBar />}
      {!processing && (
        <>
          <TitleContainer>
            <TitlePrefix>
              {intl.formatMessage(messages.headerPrefix)}
            </TitlePrefix>
            <TitleSuffix>
              {intl.formatMessage(messages.headerSuffix)}
            </TitleSuffix>
          </TitleContainer>
          <AmountContainer>
            <AmountLabel>
              {intl.formatMessage(messages.amountLabel)}
            </AmountLabel>
            <AmountValue>{`${order.currency} ${order.amount}`}</AmountValue>
          </AmountContainer>
          <OptionsContainer>
            <Option
              intl={intl}
              testid="option-online-payment"
              text={messages.onlinePayment}
              onClick={handleOnlinePayment}
              icon={logoOnlinePayment}
            />
            <Option
              intl={intl}
              testid="option-cash-on-delivery"
              text={messages.cashOnDelivery}
              icon={logoCashOnDelivery}
              onClick={handleCashOnDelivery}
            />
            <Option
              intl={intl}
              testid="option-card-on-delivery"
              text={messages.cardOnDelivery}
              icon={logoCardOnDelivery}
              onClick={handleCardOnDelivery}
            />
          </OptionsContainer>
        </>
      )}
    </>
  );
};

const PaymentOptions = ({ history, intl }: PaymentOptionsProps) => {
  const { orderId } = useParams();

  const [orderDetails, setOrderDetails] = useState<Order>(null);

  const redirectToOrderStatus = () => history.push(PAYMENT_STATUS_PAGE);

  const fetchOrderDetails = async (id: string) => {
    try {
      const order = await ordersApi.fetchOrderDetails(id);
      if (!validOrderStatuses.includes(order.status)) {
        redirectToOrderStatus();
        return;
      }
      setOrderDetails(order);
    } catch (err) {
      // eslint-disable-next-line prettier/prettier
      const responseStatus = err.response?.status;
      if (responseStatus !== 401 && responseStatus !== 403)
        redirectToOrderStatus();
    }
  };

  useEffect(() => {
    fetchOrderDetails(orderId);
  }, [orderId]);

  const initialized = !!orderDetails;

  return (
    <MainContainer>
      <MpcLogo src={mpcGrayLogo} />
      {!initialized && <ProgressBar />}
      {initialized && (
        <Options order={orderDetails} intl={intl} history={history} />
      )}
    </MainContainer>
  );
};

const PaymentOptionsPropTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }),
  intl: intlShape,
};

type PaymentOptionsProps = InferProps<typeof PaymentOptionsPropTypes>;

PaymentOptions.propTypes = PaymentOptionsPropTypes;

export default injectIntl(PaymentOptions);
