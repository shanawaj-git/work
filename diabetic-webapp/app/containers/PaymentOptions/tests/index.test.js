import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { fireEvent, render, wait } from 'react-testing-library';

import mockStripe from 'tests/utils/mockStripe';
import { OrderStatus, PaymentType } from 'apis/types';
import * as ordersApi from 'apis/orders';

// eslint-disable-next-line import/first
import PaymentOptions from '../index';

jest.mock('@stripe/react-stripe-js', () => mockStripe());

const mockOrderId = '1000';

const mockOrder = {
  id: mockOrderId,
  amount: 123.33,
  currency: 'AED',
  status: OrderStatus.RequirePayment,
};

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useParams: () => ({
    orderId: mockOrderId,
  }),
}));

describe('<PaymentOptions />', () => {
  // eslint-disable-next-line no-global-assign
  window = Object.assign(window, {
    visualViewport: {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    },
  });

  // eslint-disable-next-line no-unused-vars
  let mockHistory;
  let fetchOrderDetailsSpy;
  let confirmPaymentSpy;

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };
    fetchOrderDetailsSpy = jest
      .spyOn(ordersApi, 'fetchOrderDetails')
      .mockResolvedValue(mockOrder);
    confirmPaymentSpy = jest
      .spyOn(ordersApi, 'confirmPayment')
      .mockResolvedValue(mockOrder);
  });

  afterEach(() => {
    mockHistory = null;
    fetchOrderDetailsSpy.mockRestore();
    confirmPaymentSpy.mockRestore();
  });

  // eslint-disable-next-line react/prop-types
  const MockPaymentOptions = ({ locale, search = '' }) => (
    <BrowserRouter>
      <IntlProvider locale={locale}>
        <PaymentOptions history={mockHistory} location={{ search }} />
      </IntlProvider>
    </BrowserRouter>
  );

  const mockClickElement = dataTestId => {
    const button = document.querySelector(`[data-testid="${dataTestId}"]`);

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
  };

  it('should render and match the snapshot', async () => {
    const {
      container: { firstChild },
    } = render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      expect(firstChild).toMatchSnapshot();
    });
  });

  it('should render the loader while api calls are in progress', async () => {
    const {
      container: { firstChild },
    } = render(<MockPaymentOptions locale="en" />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should invoke fetchOrderDetails with the order id', async () => {
    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      expect(fetchOrderDetailsSpy).toBeCalledWith(mockOrderId);
    });
  });

  it('should redirect to payment input screen on selecting Pay online option', async () => {
    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      mockClickElement('option-online-payment');
      expect(mockHistory.push).toBeCalledWith('payments/input');
    });
  });

  it('should invoke confirmPayment api with the orderId and redirect to status page upon cash on delivery option', async () => {
    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      mockClickElement('option-cash-on-delivery');
      expect(confirmPaymentSpy).toBeCalledWith(
        mockOrderId,
        PaymentType.CashOnDelivery,
      );
      expect(mockHistory.push).toBeCalledWith('payments/status');
    });
  });

  it('should invoke confirmPayment api with the orderId and redirect to status page upon cash on delivery option', async () => {
    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      mockClickElement('option-card-on-delivery');
      expect(confirmPaymentSpy).toBeCalledWith(
        mockOrderId,
        PaymentType.CardOnDelivery,
      );
      expect(mockHistory.push).toBeCalledWith('payments/status');
    });
  });

  it('should redirect to status page upon failure of confirmPayment api', async () => {
    confirmPaymentSpy.mockRejectedValue(new Error('Some error'));
    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      mockClickElement('option-cash-on-delivery');
      expect(confirmPaymentSpy).toBeCalledWith(
        mockOrderId,
        PaymentType.CashOnDelivery,
      );
      expect(mockHistory.push).toBeCalledWith('payments/status');
    });
  });

  it('should redirect to the payment status page if the order status is not valid', async () => {
    fetchOrderDetailsSpy.mockResolvedValue({
      ...mockOrder,
      status: OrderStatus.Pending,
    });

    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      expect(mockHistory.push).toBeCalledWith('payments/status');
    });
  });

  it('should redirect to payment status page when fetchOrderDetails is failed', async () => {
    fetchOrderDetailsSpy.mockRejectedValue(new Error('Some error.'));

    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      expect(mockHistory.push).toBeCalledWith('payments/status');
    });
  });

  it('should not redirect to status page on Unauthorized error on fetching order details', async () => {
    fetchOrderDetailsSpy.mockRejectedValue({ response: { status: 401 } });

    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      expect(mockHistory.push).not.toBeCalled();
    });
  });

  it('should not redirect to status page on Forbidden error on fetching order details', async () => {
    fetchOrderDetailsSpy.mockRejectedValue({ response: { status: 403 } });

    render(<MockPaymentOptions locale="en" />);
    await wait(() => {
      expect(mockHistory.push).not.toBeCalled();
    });
  });
});
