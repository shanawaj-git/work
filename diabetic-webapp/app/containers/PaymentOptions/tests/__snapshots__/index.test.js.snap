// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaymentOptions /> should render and match the snapshot 1`] = `
<div
  class="sc-bdVaJa eQpvyC"
>
  <img
    class="sc-htpNat icrQpi"
    src="IMAGE_MOCK"
  />
  <div
    class="sc-bxivhb dpdSOk"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 sc-ifAKCX cHadpr css-1reifv3-MuiTypography-root"
    >
      Select your
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 sc-EHOje bAcCwx css-1reifv3-MuiTypography-root"
    >
      payment method
    </p>
  </div>
  <div
    class="sc-bZQynM dUmxn"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 sc-gzVnrw ccLiOJ css-1reifv3-MuiTypography-root"
    >
      Total to pay
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 sc-htoDjs hzhiNv css-1reifv3-MuiTypography-root"
    >
      AED 123.33
    </p>
  </div>
  <div
    class="sc-dnqmqq kbYuMr"
  >
    <button
      class="root StyledButton-sc-1apep4b-0 hPQGaD MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="option-online-payment"
      tabindex="0"
      type="button"
    >
      <div
        style="visibility: visible;"
      >
        <div
          class="sc-iwsKbI fNnfsR"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 sc-gZMcBi giNnHm css-1reifv3-MuiTypography-root"
          >
            Pay online
          </p>
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
          >
            <img
              alt="[object Object]"
              class="option-icon"
              src="IMAGE_MOCK"
            />
          </span>
        </div>
      </div>
    </button>
    <button
      class="root StyledButton-sc-1apep4b-0 hPQGaD MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="option-cash-on-delivery"
      tabindex="0"
      type="button"
    >
      <div
        style="visibility: visible;"
      >
        <div
          class="sc-iwsKbI fNnfsR"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 sc-gZMcBi giNnHm css-1reifv3-MuiTypography-root"
          >
            Cash on delivery
          </p>
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
          >
            <img
              alt="[object Object]"
              class="option-icon"
              src="IMAGE_MOCK"
            />
          </span>
        </div>
      </div>
    </button>
    <button
      class="root StyledButton-sc-1apep4b-0 hPQGaD MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="option-card-on-delivery"
      tabindex="0"
      type="button"
    >
      <div
        style="visibility: visible;"
      >
        <div
          class="sc-iwsKbI fNnfsR"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 sc-gZMcBi giNnHm css-1reifv3-MuiTypography-root"
          >
            Card on delivery
          </p>
          <span
            aria-hidden="true"
            class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-kp9ftd-MuiIcon-root"
          >
            <img
              alt="[object Object]"
              class="option-icon"
              src="IMAGE_MOCK"
            />
          </span>
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PaymentOptions /> should render the loader while api calls are in progress 1`] = `
<div
  class="sc-bdVaJa eQpvyC"
>
  <img
    class="sc-htpNat icrQpi"
    src="IMAGE_MOCK"
  />
  <div
    class="sc-bwzfXH eOoZbk"
  >
    <div
      class="MuiBox-root css-15t75f3"
    >
      <span
        aria-valuenow="100"
        class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorPrimary css-1e5kc5y-MuiCircularProgress-root"
        role="progressbar"
        style="width: 90px; height: 90px; transform: rotate(-90deg);"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="21"
            stroke-width="2"
            style="stroke-dasharray: 131.947; stroke-dashoffset: 0.000px;"
          />
        </svg>
      </span>
      <svg
        style="position: absolute;"
      >
        <defs>
          <lineargradient
            id="my_gradient"
          >
            <stop
              offset="0%"
              stop-color="#00664f"
            />
            <stop
              offset="100%"
              stop-color="#c5c5c5"
            />
          </lineargradient>
        </defs>
      </svg>
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1utiupp-MuiCircularProgress-root"
        role="progressbar"
        style="width: 90px; height: 90px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate MuiCircularProgress-circleDisableShrink css-79nvmn-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="21"
            stroke-width="2"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
